dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/programrule/implementers/event/AssignValueExecutorTest.java

```diff
@@ -0,0 +1,354 @@
+
+    private final static String DATAELEMENT_OLD_VALUE = "10.0";
+
+    private final static String DATAELEMENT_NEW_VALUE = "24.0";
+
+    private static ProgramStage firstProgramStage;
+
+    private static ProgramStage secondProgramStage;
+
+    private static DataElement dataElementA;
+
+    private static DataElement dataElementB;
+
+    private TrackerBundle bundle;
+
+    @Mock
+    private TrackerPreheat preheat;
+
+    @Mock
+    private SystemSettingManager systemSettingManager;
+
+    @BeforeEach
+    void setUpTest()
+    {
+        firstProgramStage = createProgramStage( 'A', 0 );
+        firstProgramStage.setValidationStrategy( ValidationStrategy.ON_UPDATE_AND_INSERT );
+        dataElementA = createDataElement( 'A' );
+        dataElementA.setUid( DATA_ELEMENT_ID );
+        dataElementA.setCode( DATA_ELEMENT_CODE );
+        ProgramStageDataElement programStageDataElementA = createProgramStageDataElement( firstProgramStage,
+            dataElementA, 0 );
+        firstProgramStage.setProgramStageDataElements( Sets.newHashSet( programStageDataElementA ) );
+        secondProgramStage = createProgramStage( 'B', 0 );
+        secondProgramStage.setValidationStrategy( ValidationStrategy.ON_UPDATE_AND_INSERT );
+        dataElementB = createDataElement( 'B' );
+        dataElementB.setUid( ANOTHER_DATA_ELEMENT_ID );
+        ProgramStageDataElement programStageDataElementB = createProgramStageDataElement( secondProgramStage,
+            dataElementB, 0 );
+        secondProgramStage.setProgramStageDataElements( Sets.newHashSet( programStageDataElementB ) );
+        when( preheat.getProgramStage( MetadataIdentifier.ofUid( firstProgramStage ) ) )
+            .thenReturn( firstProgramStage );
+        when( preheat.getProgramStage( MetadataIdentifier.ofUid( secondProgramStage ) ) )
+            .thenReturn( secondProgramStage );
+        when( preheat.getDataElement( DATA_ELEMENT_ID ) ).thenReturn( dataElementA );
+        bundle = TrackerBundle.builder().build();
+        bundle.setPreheat( preheat );
+        when( systemSettingManager.getBooleanSetting( SettingKey.RULE_ENGINE_ASSIGN_OVERWRITE ) )
+            .thenReturn( Boolean.FALSE );
+    }
+
+    @Test
+    void shouldAssignDataValueValueForEventsWhenDataValueIsEmpty()
+    {
+        when( preheat.getIdSchemes() ).thenReturn( TrackerIdSchemeParams.builder().build() );
+        Event eventWithDataValueNOTSet = getEventWithDataValueNOTSet();
+        List<Event> events = List.of( eventWithDataValueNOTSet );
+        bundle.setEvents( events );
+
+        AssignValueExecutor executor = new AssignValueExecutor( systemSettingManager,
+            "", DATAELEMENT_NEW_VALUE, DATA_ELEMENT_ID, eventWithDataValueNOTSet.getDataValues() );
+
+        Optional<ProgramRuleIssue> warning = executor.executeRuleAction( bundle, eventWithDataValueNOTSet );
+
+        Event Event = bundle.getEvents().stream().filter( e -> e.getEvent().equals( SECOND_EVENT_ID ) ).findAny().get();
```

Comment:
Is this streaming/filtering very similar in all the tests? can you extract a helper?